### PR TITLE
Added steadystate_floquet function

### PR DESCRIPTION
**Pull request by albertomercurio**
**Description**
The current steadystate solver requires a time-independent Hamiltonian. This is obvious because a time-dependent Hamiltonian cannot have a steadystate.
However, if the Hamiltonian can be decomposed in a time-independent part H_0 and a periodic time-dependent part cos(w t) * H_i, the system reaches a stroboscopic steadystate. If the perturbation is small, this stroboscopic steadystate fluctuates so slightly that it can be considered as a real steadystate.

I rewrote the Matrix Continued Fractions method [implemented in the Quantum Toolbox for Matlab](https://copilot.caltech.edu/documents/16743/qousersguide.pdf) to fit with python and qutip objects. It is 100% working, with running times lower than the second with an Hamiltonian 40 x 40.

[Here](https://ibb.co/PthSzRd) the plot showing the comparison between the master equation dynamics, and the steadystate solution.

Currently, there is the possibility to solve only this kind of problems. In the future we can add the support for multiple operators with different frequencies and amplitudes.

**Related issues or PRs**
issue #1659

**Changelog**
Added the function steadystate_floquet, which returns the "effective" steadystate of a periodic driven system.

### CODE REVIEW

**Comment by Ericgig** [comment]
Hi albertomercurio, 
Thank you for the contribution.

We will need at least one automated test: a function in `qutip/tests/test_correlation.py` that call `steadystate_floquet` and would fail if something is wrong. For example, in the next major version of qutip, `Qobj.data.tocsc()` will fail. We want it to be caught when running the tests in `qutip/tests`  instead of having to remember to update it.

It could be simply to running your manual test and instead of plotting the result, checking it with `numpy.testing.assert_allclose`. Ideally it should be done twice, once for `sparse=True` and once for `sparse=False`.

**Comment by albertomercurio** [comment]
I added the test in `qutip/tests/test_correlation.py` and pytest returns no errors. Why not put it inside the `qutip/tests/test_steadystate.py` file instead?

**Comment by Ericgig** [comment]
Yes, sorry, it should be in `test_steadystate.py`.

**Comment by albertomercurio** [comment]
Ok, I just put it inside the `qutip/tests/test_steadystae.py` file.

File: qutip/steadystate.py [hunk 1]
```diff
@@ -914,6 +914,92 @@ def _iter_count(r):
...
+        L_p_array = L_p.todense()
+        L_m_array = L_m.todense()
+
+        Id = sp.eye(N ** 2, format="csc", dtype=np.complex128)
+        S = T = sp.csc_matrix((N ** 2, N ** 2), dtype=np.complex128)
+
+        for n_i in np.arange(nmax, 0, -1):
+            L = sp.csc_matrix(L_0 - 1j * n_i * w_d * Id + L_m.dot(S))
+            L.sort_indices()
+            LU = splu(L)
```
Comment (Ericgig) [inline qutip/steadystate.py:969]: Have you tired other choice of solver `splu` vs `spsolve` or `gmres`, etc.
`steadystate` allows the choice of solver and method, while they are not all great, the option exist. 
Does it make sense to have something similar here?

File: qutip/steadystate.py [hunk 2]
```diff
@@ -914,6 +914,92 @@ def _iter_count(r):
...
+    c_ops : list
+        A list of collapse operators.
+
+    Op_t : :obj:`~Qobj`
+        The the interaction operator which is multiplied by the cosine
+
+    w_d : float, default 1.0
+        The frequency of the drive
+
+    nmax : int, default 3
```
Comment (Ericgig) [inline qutip/steadystate.py:941]: This is the number of iteration, not the maximum.
Why is the default 3? How to know how many is enough?

**Review by Ericgig — COMMENTED** [review]
Overall looks good.

File: qutip/steadystate.py [hunk 3]
```diff
@@ -914,6 +914,92 @@ def _iter_count(r):
...
+    dm : qobj
+        Steady state density matrix.
+    """
+    if sparse:
+        N = H_0.shape[0]
+
+        L_0 = liouvillian(H_0, c_ops).data.tocsc()
+        L_t = liouvillian(Op_t)
+        L_p = (0.5 * L_t).data.tocsc()
+        L_m = L_p  # (0.5 * A_l * L_t).data
```
[the remark this answers is not in the document]
Comment (albertomercurio) [reply qutip/steadystate.py:959]: I removed the unnecessary comment. Here L_p and L_m are related to the positive and negative frequency respectively. In the special case of a cosinusoidal drive they are the same. In the future, they could also be different. For this version should I specify only one operator (for example L_d = L_p = L_m) ?

File: qutip/steadystate.py [hunk 4]
```diff
@@ -914,6 +914,92 @@ def _iter_count(r):
...
+        The maximum number of iteration for the solver
+
+    sparse : bool, default False
+        Solve for the steady state using sparse algorithms.
+        Actually, dense seems to be faster.
+
+    Returns
+    -------
+    dm : qobj
+        Steady state density matrix.
```
[the remark this answers is not in the document]
Comment (albertomercurio) [reply qutip/steadystate.py:951]: Perfect, how should I add it?

File: qutip/steadystate.py (hunk 1 above)
Comment (albertomercurio) [reply qutip/steadystate.py:969]: I have tried `spsolve` and it is much slower than `splu`, I have tried it for different type of systems. For `gmres`, as far as I know it is only valid for systems of the type `Ax = b` where `b` is a vector. Here `b` is a matrix. 
Thus, I think the only possible options are `splu` (faster) and `spsolve` (slower).

File: qutip/steadystate.py (hunk 2 above)
Comment (albertomercurio) [reply qutip/steadystate.py:941]: You are right, this is the number of iterations. When the steadystate does not change anymore increasing this number it reaches the convergence. I tried it for different systems and n = 3 seems to be a valid choice to balance resolution and speed.

File: qutip/steadystate.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Ericgig) [reply qutip/steadystate.py:951]: Inside a note at the end of the doc-string, something like:
```
.. note::
    See: Sze Meng Tan, https://copilot.caltech.edu/documents/16743/qousersguide.pdf, Section (10.16)
```

[the remark this answers is not in the document]
Comment (albertomercurio) [reply qutip/steadystate.py:951]: Perfect!

**Review by Ericgig — COMMENTED** [review]
Let's leave a comment on why both `L_p` and `L_m` are used.
Other than that, I am happy with it.

We are working on major changes in Qutip that will break `steadystate_floquet`. Would you be interested in porting this to the new version eventually?

**Comment by albertomercurio** [comment]
> We are working on major changes in Qutip that will break `steadystate_floquet`. Would you be interested in porting this to the new version eventually?

It could be great!

**Review by Ericgig — APPROVED** [review]

**Comment by parsidd** [comment]
Hi!
I am interested in using a floquet based steady state solver (with time-independent Lindblad dissipators) approach in qutip, and came across this conversation. Am I right in understanding that the latest Qutip version does not have a steady state solver for the floquet method, or am I just missing something? 

I noticed that there does exist a floquet master equation solver, but it seems to only allow for one dissipator, and I need more than that to simulate my system.

**Comment by Ericgig** [comment]
The latest version has [`steadystate_floquet`](https://qutip.readthedocs.io/en/qutip-5.0.x/apidoc/functions.html#qutip.solver.steadystate.steadystate_floquet). This PR added it.

The floquet solver support a list of `c_ops`. The note in the documentation was not removed when the feature was added.

**Comment by parsidd** [comment]
Thanks a lot @Ericgig ! I somehow missed that function in the documentation

**Pull request merged**